wenet/dataset/dataset.py

```diff
@@ -98,45 +169,132 @@ def spec_augmentation(x,
-        fid.seek(0)
-        arr = fid.read().split()
-        assert (arr[0] == '[')
-        assert (arr[-2] == '0')
-        assert (arr[-1] == ']')
-        feat_dim = int((len(arr) - 2 - 2) / 2)
-        for i in range(1, feat_dim + 1):
-            means.append(float(arr[i]))
-        count = float(arr[feat_dim + 1])
-        for i in range(feat_dim + 2, 2 * feat_dim + 2):
-            variance.append(float(arr[i]))
+    This distortion will not change the length of the waveform.
 
-    for i in range(len(means)):
-        means[i] /= count
-        variance[i] = variance[i] / count - means[i] * means[i]
-        if variance[i] < 1.0e-20:
-            variance[i] = 1.0e-20
-        variance[i] = 1.0 / math.sqrt(variance[i])
-    cmvn = np.array([means, variance])
-    return cmvn
+    Args:
+        waveform: numpy float tensor, (length,)
+        distortion_methods_conf: a list of config for ditortion method.
+            a method will be randomly selected by 'method_rate' and
+            apply on the waveform.
+
+    Returns:
+        distorted waveform.
+    """
+    r = random.uniform(0, 1)
+    acc = 0.0
+    for distortion_method in distortion_methods_conf:
+        method_rate = distortion_method['method_rate']
+        acc += method_rate
+        if r < acc:
+            distortion_type = distortion_method['name']
+            distortion_conf = distortion_method['params']
+            point_rate = distortion_method['point_rate']
+            return distort_wav_conf(waveform, distortion_type,
+                                    distortion_conf , point_rate)
+    return waveform
 
+# add speed perturb when loading wav
+# return augmented, sr
+def _load_wav_with_speed(wav_file, speed):
+    """ Load the wave from file and apply speed perpturbation
 
-def _load_from_file(batch):
+    Args:
+        wav_file: input feature, T * F 2D
+
+    Returns:
+        augmented feature
+    """
+    if speed == 1.0:
+        return torchaudio.load_wav(wav_file)
+    else:
+        si, _ = torchaudio.info(wav_file)
+        E = torchaudio.sox_effects.SoxEffectsChain()
+        E.append_effect_to_chain('speed', speed)
+        E.append_effect_to_chain("rate", si.rate)
+        E.set_input_file(wav_file)
+        return E.sox_build_flow_effects()
+
+
+
+def _extract_feature(batch, speed_perturb, wav_distortion_conf,
+                     feature_extraction_conf):
+    """ Extract acoustic fbank feature from origin waveform.
+
+    Speed perturbation and wave amplitude distortion is optional.
+
+    Args:
+        batch: a list of tuple (wav id , wave path).
+        speed_perturb: bool, whether or not to use speed pertubation.
+        wav_distortion_conf: a dict , the config of wave amplitude distortion.
+        feature_extraction_conf:a dict , the config of fbank extraction.
+
+    Returns:
+        (keys, feats, labels)
+    """
+    keys = []
+    feats = []
+    lengths = []
+    wav_dither = wav_distortion_conf['wav_dither'] / 32768.0
```

Comment:
why dither is computed in this way?

Reply:
The value of torchaudio wavefrom is in [-1,1] while kaldi value is in [-32768,32767].